---
title: Preprocessing and normalization of single-cell RNA-seq read counts
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
bibliography: ref.bib
---   

```{r, echo=FALSE, message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
set.seed(651)
```

This workflow is adapted from portions of the [simpleSingleCell Bioconductor workflow](https://www.bioconductor.org/help/workflows/simpleSingleCell/) by Aaron Lun, Davis McCarthy, and John Marioni. Some sections have been simplified, additional material has been added exploring additional methods, and exercises have been added.

# Overview

In this workflow, we use a relatively simple dataset [@lun2017assessing] to introduce most of the concepts of scRNA-seq data analysis.
This dataset contains two plates of 416B cells (an immortalized mouse myeloid progenitor cell line), processed using the Smart-seq2 protocol [@picelli2014fulllength].
A constant amount of spike-in RNA from the External RNA Controls Consortium (ERCC) was also added to each cell's lysate prior to library preparation.
High-throughput sequencing was performed and the expression of each gene was quantified by counting the total number of reads mapped to its exonic regions.
Similarly, the quantity of each spike-in transcript was measured by counting the number of reads mapped to the spike-in reference sequences.

Counts for all genes/transcripts in each cell are available from ArrayExpress using the accession number [E-MTAB-5522](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5522).
We download both the count tables (in the "processed files") as well as the metadata file using the `r Biocpkg("BiocFileCache")` package. 
This saves the files to a local cache (`raw_data`) and avoids re-downloading them if they are already present.

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("../raw_data", ask = FALSE)
lun.zip <- bfcrpath(bfc, 
    file.path("https://www.ebi.ac.uk/arrayexpress/files",
        "E-MTAB-5522/E-MTAB-5522.processed.1.zip"))
lun.sdrf <- bfcrpath(bfc, 
    file.path("https://www.ebi.ac.uk/arrayexpress/files",
        "E-MTAB-5522/E-MTAB-5522.sdrf.txt"))
unzip(lun.zip, exdir = "../raw_data")
```


# Setting up the data

## Loading in the count matrix

Our first task is to load the count matrices into memory.
One matrix was generated for each plate of cells used in the study.
In each matrix, each row represents an endogenous gene or a spike-in transcript, and each column represents a cell.
Subsequently, the count in each entry of the matrix represents the number of reads mapped to a particular gene/transcript in a particular cell.

```{r}
plate1 <- read.delim(file.path("../raw_data/counts_Calero_20160113.tsv"), 
    header=TRUE, row.names=1, check.names=FALSE)
plate2 <- read.delim(file.path("../raw_data/counts_Calero_20160325.tsv"), 
    header=TRUE, row.names=1, check.names=FALSE)
gene.lengths <- plate1$Length # First column is the gene length.
plate1 <- as.matrix(plate1[,-1]) # Discarding gene length (as it is not a cell).
plate2 <- as.matrix(plate2[,-1])
rbind(Plate1=dim(plate1), Plate2=dim(plate2))
```

We combine the two matrices into a single object for further processing.
This is done after verifying that the genes are in the same order between the two matrices.

```{r}
stopifnot(identical(rownames(plate1), rownames(plate2)))
all.counts <- cbind(plate1, plate2)
```

Next, we'll store the count matrix in a `SingleCellExperiment` object from the `r Biocpkg("SingleCellExperiment")` package.
This allows different types of row- and column-level metadata to be stored alongside the counts for synchronized manipulation throughout the workflow.

```{r}
library(SingleCellExperiment)
sce <- SingleCellExperiment(list(counts=all.counts))
rowData(sce)$GeneLength <- gene.lengths
sce
```

We identify the rows corresponding to ERCC spike-in transcripts from the row names.
We store this information in the `SingleCellExperiment` object for future use.
This is necessary as spike-ins require special treatment in downstream steps such as normalization.

```{r}
isSpike(sce, "ERCC") <- grepl("^ERCC", rownames(sce))
summary(isSpike(sce, "ERCC"))
```

This dataset is slightly unusual in that it contains information from another set of spike-in transcripts, the Spike-In RNA Variants (SIRV) set.
For simplicity, we will only use the ERCC spike-ins in this analysis.
Thus, we must remove the rows corresponding to the SIRV transcripts prior to further analysis, which can be done simply by subsetting the `SingleCellExperiment` object.

```{r}
is.sirv <- grepl("^SIRV", rownames(sce))
sce <- sce[!is.sirv,] 
summary(is.sirv)
```

## Incorporating cell-based annotation

We load in the metadata for each library/cell from the `sdrf.txt` file.
It is important to check that the rows of the metadata table are in the same order as the columns of the count matrix.
Otherwise, incorrect metadata will be assigned to each cell.

```{r}
metadata <- read.delim(lun.sdrf, check.names=FALSE, header=TRUE)
m <- match(colnames(sce), metadata[["Source Name"]]) # Enforcing identical order.
stopifnot(all(!is.na(m))) # Checking that nothing's missing.
metadata <- metadata[m,]
head(colnames(metadata))
```

We only retain relevant metadata fields to avoid storing unnecessary information in the `colData` of the `SingleCellExperiment` object.
In particular, we keep the plate of origin (i.e., `block`) and phenotype of each cell.
The second field is relevant as all of the cells contain a _CBFB-MYH11_ oncogene, but the expression of this oncogene is only induced in a subset of the cells.

```{r}
colData(sce)$Plate <- factor(metadata[["Factor Value[block]"]])
pheno <- metadata[["Factor Value[phenotype]"]]
levels(pheno) <- c("induced", "control")
colData(sce)$Oncogene <- pheno
```

## Incorporating gene-based annotation

Feature-counting tools typically report genes in terms of standard identifiers from Ensembl or Entrez.
These identifiers are used as they are unambiguous and highly stable.
However, they are difficult to interpret compared to the gene symbols which are more commonly used in the literature.
Given the Ensembl identifiers, we obtain the corresponding gene symbols using annotation packages like `r Biocpkg("org.Mm.eg.db")`.

```{r}
library(org.Mm.eg.db)
symb <- mapIds(org.Mm.eg.db, keys=rownames(sce), keytype="ENSEMBL", column="SYMBOL")
rowData(sce)$ENSEMBL <- rownames(sce)
rowData(sce)$SYMBOL <- symb
head(rowData(sce))
```

It is often desirable to rename the row names of `sce` to the gene symbols, as these are easier to interpret.
However, this requires some work to account for missing and duplicate symbols.
The code below will replace missing symbols with the Ensembl identifier and concatenate duplicated symbols with the (unique) Ensembl identifiers.

```{r}
library(scater)
rownames(sce) <- uniquifyFeatureNames(rowData(sce)$ENSEMBL, rowData(sce)$SYMBOL)
head(rownames(sce))
```

We also determine the chromosomal location for each gene using the `r Biocpkg("TxDb.Mmusculus.UCSC.mm10.ensGene")` package.
This will be useful later as several quality control metrics will be computed from rows corresponding to mitochondrial genes.

```{r}
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
location <- mapIds(TxDb.Mmusculus.UCSC.mm10.ensGene, keys=rowData(sce)$ENSEMBL, 
    column="CDSCHROM", keytype="GENEID")
rowData(sce)$CHR <- location
```

# Quality control on the cells 

## Defining the quality control metrics

Low-quality cells need to be removed to ensure that technical effects do not distort downstream analysis results.
We use several quality control (QC) metrics:

- The library size is defined as the total sum of counts across all features, i.e., genes and spike-in transcripts.
Cells with small library sizes are of low quality as the RNA has not been efficiently captured (i.e., converted into cDNA and amplified) during library preparation.
- The number of expressed features in each cell is defined as the number of features with non-zero counts for that cell.
Any cell with very few expressed genes is likely to be of poor quality as the diverse transcript population has not been successfully captured.
- The proportion of reads mapped to spike-in transcripts is calculated relative to the library size for each cell.
High proportions are indicative of poor-quality cells, where endogenous RNA has been lost during processing (e.g., due to cell lysis or RNA degradation).
The same amount of spike-in RNA to each cell, so an enrichment in spike-in counts is symptomatic of loss of endogenous RNA.
- In the absence of spike-in transcripts, the proportion of reads mapped to genes in the mitochondrial genome can also be used.
High proportions are indicative of poor-quality cells [@islam2014quantitative;@ilicic2016classification], possibly because of loss of cytoplasmic RNA from perforated cells.
The reasoning is that mitochondria are larger than individual transcript molecules and less likely to escape through tears in the cell membrane.

For each cell, we calculate these quality control metrics using the `calculateQCMetrics` function from the `r Biocpkg("scater")` package [@mccarthy2017scater].
These are stored in the row- and column-wise metadata of the `SingleCellExperiment` for future reference.

```{r}
mito <- which(rowData(sce)$CHR=="chrM")
sce <- calculateQCMetrics(sce, feature_controls=list(Mt=mito))
colnames(colData(sce))
```

The distributions of these metrics are shown in Figure \@ref(fig:qcplot416b), stratified by oncogene induction status and plate of origin.
The aim is to remove putative low-quality cells that have **low library sizes**, **low numbers of expressed features**, and **high spike-in (or mitochondrial) proportions**.
Such cells can interfere with downstream analyses, e.g., by forming distinct clusters that complicate interpretation of the results.

```{r qcplot416b, fig.wide=TRUE, fig.cap="Distributions of various QC metrics for all cells in the 416B dataset. This includes the library sizes, number of expressed genes, and proportion of reads mapped to spike-in transcripts or mitochondrial genes."}
sce$PlateOnco <- paste0(sce$Oncogene, ".", sce$Plate)
multiplot(
    plotColData(sce, y="total_counts", x="PlateOnco"),
    plotColData(sce, y="total_features_by_counts", x="PlateOnco"),
    plotColData(sce, y="pct_counts_ERCC", x="PlateOnco"),
    plotColData(sce, y="pct_counts_Mt", x="PlateOnco"),
    cols=2)
```


## Identifying outliers for each metric 

Picking a threshold for these metrics is not straightforward as their absolute values depend on the experimental protocol.
For example, sequencing to greater depth will lead to more reads and more expressed features, regardless of the quality of the cells.
Similarly, using more spike-in RNA in the protocol will result in higher spike-in proportions.
To obtain an adaptive threshold, we assume that most of the dataset consists of high-quality cells, and identify cells that are outliers for the various QC metrics.

Outliers are defined based on the median absolute deviation (MADs) from the median value of each metric across all cells.
We remove cells with log-library sizes that are more than 3 MADs below the median log-library size.
A log-transformation improves resolution at small values, especially when the MAD of the raw values is comparable to or greater than the median.
We also remove cells where the log-transformed number of expressed genes is 3 MADs below the median value.

```{r}
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", 
    log=TRUE, batch=sce$PlateOnco)
feature.drop <- isOutlier(sce$total_features_by_counts, nmads=3, type="lower", 
    log=TRUE, batch=sce$PlateOnco)

sum(libsize.drop)
sum(feature.drop)
```

The `batch=` argument ensures that outliers are identified _within_ each level of the specified plate/oncogene factor.
This allows `isOutlier()` to accommodate systematic differences in the QC metrics across plates (Figure \@ref(fig:qcplot416b)),
which can arise due to technical differences in processing (e.g., differences in sequencing depth) rather than any changes in quality.
The same reasoning applies to the oncogene induction status, where induced cells may have naturally fewer expressed genes for biological reasons.
Failing to account for these systematic differences would inflate the MAD estimate and compromise the removal of low-quality cells.

We identify outliers for the proportion-based metrics in a similar manner.
Here, no transformation is required as we are identifying large outliers, for which the distinction should be fairly clear on the raw scale.
We do not need to use the mitochondrial proportions as we already have the spike-in proportions (which serve a similar purpose) for this dataset.
This avoids potential issues arising from genuine differences in mitochondrial content between cell types that may confound outlier identification.

```{r}
spike.drop <- isOutlier(sce$pct_counts_ERCC, nmads=3, type="higher",
    batch=sce$PlateOnco)
sum(spike.drop)
```

Subsetting by column will retain only the high-quality cells that pass each filter described above.
We examine the number of cells removed by each filter as well as the total number of retained cells.
Removal of a substantial proportion of cells (> 10%) may be indicative of an overall issue with data quality.

```{r}
keep <- !(libsize.drop | feature.drop | spike.drop)
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop),
    BySpike=sum(spike.drop), Remaining=sum(keep))
```

We then subset the `SingleCellExperiment` object to retain only the putative high-quality cells.
We also save the original object to file for later use.

```{r}
sce$PassQC <- keep
saveRDS(sce, file="416B_preQC.rds")
sce <- sce[,keep]
dim(sce)
```


# Examining gene-level expression metrics

We examine the identities of the most highly expressed genes (Figure \@ref(fig:topgene416b)).
This should generally be dominated by constitutively expressed transcripts, such as those for ribosomal or mitochondrial proteins. Here, for example, we see that the mitochondrial proteins CYTB and COX1 are some of the most highly expressed genes.
The presence of other classes of features may be cause for concern if they are not consistent with expected biology.
For example, a top set containing many spike-in transcripts suggests that too much spike-in RNA was added during library preparation, while the absence of ribosomal proteins and/or the presence of their pseudogenes are indicative of suboptimal alignment.

```{r topgene416b, fig.wide=TRUE, fig.cap="Percentage of total counts assigned to the top 25 most highly-abundant features in the 416B dataset. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells. Bars are coloured by the total number of expressed features in each cell, while circles are coloured according to whether the feature is labelled as a control feature."}
plotHighestExprs(sce, n=25)  + theme(text = element_text(size=14))
```

Another useful metric is the average count for each gene, computed across all cells in the dataset.
We calculate this using the `calcAverage()` function, which also performs adjustment for library size differences between cells. 
We typically observe a peak of moderately expressed genes following a plateau of lowly expressed genes (Figure \@ref(fig:abhist416b)).
Note that the genes that are not detected in any cell are not shown in the plot since 
here no pseudocount is added.

```{r abhist416b, fig.cap="Histogram of log-average counts for all genes in the 416B dataset."}
ave.counts <- calcAverage(sce, use_size_factors=FALSE)
hist(log10(ave.counts), breaks=100, main="", col="grey80", 
    xlab=expression(Log[10]~"average count"))
sum(ave.counts == 0)
```

We also examine the number of cells that express each gene.
This is closely related to the average count for most genes, as expression in many cells will result in a higher average (Figure \@ref(fig:nexprshist416b)).
Genes expressed in very few cells are often uninteresting as they are driven by amplification artifacts (though they may also also arise from rare populations).

```{r nexprshist416b, fig.cap="The number of cells expressing each gene in the 416B dataset, plotted against the log-average count. Intensity of colour corresponds to the number of genes at any given location."}
num.cells <- nexprs(sce, byrow=TRUE)
smoothScatter(log10(ave.counts), num.cells, ylab="Number of cells", 
    xlab=expression(Log[10]~"average count"))
```

We remove genes that are not expressed in any cell to avoid unnecessary computational work in downstream steps. 
Further filtering may be applied on a step-by-step basis within specific functions - 
see comments `r simpleSingleCell:::.link("misc", "comments on filtering by abundance", "here")` for a more detailed discussion.

```{r}
to.keep <- num.cells > 0
sce <- sce[to.keep,]
summary(to.keep)
```

# Normalization of cell-specific biases

## Using the deconvolution method to deal with zero counts

Read counts are subject to differences in capture efficiency and sequencing depth between cells [@stegle2015computational].
Normalization is required to eliminate these cell-specific biases prior to downstream quantitative analyses.
This is often done by assuming that most genes are not differentially expressed (DE) between cells.
Any systematic difference in count size across the non-DE majority of genes between two cells is assumed to represent bias and is removed by scaling.
More specifically, "size factors" are calculated that represent the extent to which counts should be scaled in each library.

Size factors can be computed with several different approaches, e.g., using the `estimateSizeFactorsForMatrix` function in the `r Biocpkg("DESeq2")` package [@anders2010differential;@love2014moderated], or with the `calcNormFactors` function [@robinson2010scaling] in the `r Biocpkg("edgeR")` package.
However, single-cell data can be problematic for these bulk data-based methods due to the dominance of low and zero counts.

To overcome this, we pool counts from many cells to increase the size of the counts for accurate size factor estimation [@lun2016pooling].
Pool-based size factors are then "deconvolved" into cell-based factors for normalization of each cell's expression profile.

```{r}
library(scran)
sce <- computeSumFactors(sce)
summary(sizeFactors(sce))
```

The size factors are well-correlated with the library sizes for all cells (Figure \@ref(fig:normplot416b)).
This suggests that most of the systematic differences between cells are driven by differences in capture efficiency or sequencing depth.
Any DE between cells would yield a non-linear trend between the total count and size factor, and/or increased scatter around the trend.
We observe some evidence of this after oncogene induction, where the size factors after induction are systematically lower.
This is consistent with composition biases [@robinson2010scaling] introduced by upregulation of genes after induction.

```{r normplot416b, fig.cap="Size factors from deconvolution, plotted against library sizes for all cells in the 416B dataset. Axes are shown on a log-scale. Wild-type cells are shown in black and oncogene-induced cells are shown in red."}
plot(sce$total_counts/1e6, sizeFactors(sce), log="xy",
    xlab="Library size (millions)", ylab="Size factor",
    col=c("red", "black")[sce$Oncogene], pch=16)
legend("bottomright", col=c("red", "black"), pch=16, cex=1.2,
    legend=levels(sce$Oncogene))
```


## Computing separate size factors for spike-in transcripts

Size factors computed from the counts for endogenous genes are usually not appropriate for normalizing the counts for spike-in transcripts.
Consider an experiment without library quantification, i.e., the amount of cDNA from each library is _not_ equalized prior to pooling and multiplexed sequencing.
Here, cells containing more RNA have greater counts for endogenous genes and thus larger size factors to scale down those counts.
However, the same amount of spike-in RNA is added to each cell during library preparation.
This means that the counts for spike-in transcripts are not subject to the effects of RNA content.
Attempting to normalize the spike-in counts with the gene-based size factors will lead to over-normalization and incorrect quantification of expression.
Similar reasoning applies in cases where library quantification is performed. 
For a constant total amount of cDNA, any increases in endogenous RNA content will suppress the coverage of spike-in transcripts.
As a result, the bias in the spike-in counts will be opposite to that captured by the gene-based size factor.

To ensure normalization is performed correctly, we compute a separate set of size factors for the spike-in set.
For each cell, the spike-in-specific size factor is defined as the total count across all transcripts in the spike-in set.
This assumes that none of the spike-in transcripts are differentially expressed, which is reasonable given that the same amount and composition of spike-in RNA should have been added to each cell [@lun2017assessing].
(See below for a more detailed discussion on spike-in normalization.)
These size factors are stored in a separate field of the `SingleCellExperiment` object by setting `general.use=FALSE` in `computeSpikeFactors`.
This ensures that they will only be used with the spike-in transcripts but not the endogenous genes.

```{r}
sce <- computeSpikeFactors(sce, type="ERCC", general.use=FALSE)
```

## Applying the size factors to normalize gene expression

The count data are used to compute normalized log-expression values for use in downstream analyses.
Each value is defined as the log~2~-ratio of each count to the size factor for the corresponding cell, after adding a prior count of 1 to avoid undefined values at zero counts.
Division by the size factor ensures that any cell-specific biases are removed.
If spike-in-specific size factors are present in `sce`, they will be automatically applied to normalize the spike-in transcripts separately from the endogenous genes. 

```{r}
sce <- normalize(sce)
```

The log-transformation is useful as it means that any differences in the values directly represent log~2~-fold changes in expression between cells.
This is usually more relevant than the absolute differences in coverage, which need to be interpreted in the context of the overall abundance.
The log-transformation also provides some measure of variance stabilization [@law2014voom], so that high-abundance genes with large variances do not dominate downstream analyses.
The computed values are stored as an `"logcounts"` matrix in addition to the other assay elements.

## Using SCnorm to estimate the count-depth relationship

As pointed out by [@bacher2017scnorm], global size factors for all genes may be problematic 
if gene counts exhibit varying relationships with read depth (referred to as the 'count-depth' 
relationship). In particular, genes with low average expression tend to be 'over-normalized'.

Instead, we can try computing cell- and gene-specific size factors using `r Biocpkg("SCnorm")`. 
First, we examine the count-depth relationship in raw data. For the sake of computational
efficiency, we'll only work with a random subset of 2000 genes in this section.

```{r, fig.height = 4, fig.width = 8}
# random subset of 2000 genes
set.seed(9)
ix <- sample(seq_len(nrow(sce)), 2000, replace=FALSE)

library(SCnorm)
countDeptEst <- plotCountDepth(Data = sce[ix,], 
                               FilterCellProportion = .1, 
                               NCores=1)
```

Next, we examine the relationship in scran-normalized data.

```{r, fig.height = 4, fig.width = 8}
countDeptEst <- plotCountDepth(Data = sce[ix,], 
                               NormalizedData = assays(sce)$logcounts[ix,],
                               FilterCellProportion = .1, 
                               NCores=1)
```

We instead carry out normalization using SCnorm and replot the slopes of the count-depth relationship. For the sake of compuational efficiency, we set K=3 (not recommended in practice).

```{r, fig.height = 4, fig.width = 8}
sce.scnorm <- SCnorm(Data = sce[ix,],
                     Conditions = colData(sce)$PlateOnco, K=3,
                     FilterCellNum = 10, NCores=1, reportSF = TRUE)

countDeptEst <- plotCountDepth(Data = sce.scnorm, 
                               NormalizedData = normcounts(sce.scnorm),
                               FilterCellProportion = .1, 
                               NCores=1)
```


# Modelling the technical noise in gene expression

Variability in the observed expression values across genes can be driven by genuine biological heterogeneity or uninteresting technical noise. 
To distinguish between these two possibiltiies, we need to model the technical component of the variance of the expression values for each gene.
We do so using the set of spike-in transcripts, which were added in the same quantity to each cell.
Thus, the spike-in transcripts should exhibit no biological variability, i.e., any variance in their counts should be technical in origin.

We use the `trendVar()` function to fit a mean-dependent trend to the variances of the log-expression values for the spike-in transcripts.
We set `block=` to block on the plate of origin for each cell, to ensure that technical differences between plates do not inflate the variances.
Given the mean abundance of a gene, the fitted value of the trend is then used as an estimate of the technical component for that gene.
The biological component of the variance is finally calculated by subtracting the technical component from the total variance of each gene with the `decomposeVar` function.

```{r}
var.fit <- trendVar(sce, parametric=TRUE, block=sce$Plate,
    loess.args=list(span=0.3))
var.out <- decomposeVar(sce, var.fit)
head(var.out)
```

Note that negative biological components are often obtained from `decomposeVar`. 
These are intuitively meaningless as it is impossible for a gene to have total variance below technical noise.
Nonetheless, such values occur due to imprecise estimation of the total variance, especially for low numbers of cells. In the absence of spike-ins, you can set `use.spikes=FALSE` to fit a trend to the variances of the endogenous genes. Alternatively, we can create a trend based on the assumption of Poisson technical noise, as described.

We visually inspect the trend to confirm that it corresponds to the spike-in variances (Figure \@ref(fig:hvgplot416b))). 
The wave-like shape is typical of the mean-variance trend for log-expression values.
A linear increase in the variance is observed as the mean increases from zero, as larger variances are possible when the counts increase.
At very high abundances, the effect of sampling noise decreases due to the law of large numbers, resulting in a decrease in the variance.

```{r hvgplot416b, fig.cap="Variance of normalized log-expression values for each gene in the 416B dataset, plotted against the mean log-expression. The blue line represents the mean-dependent trend fitted to the variances of the spike-in transcripts (red)."}
plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", 
    ylab="Variance of log-expression")
curve(var.fit$trend(x), col="dodgerblue", lwd=2, add=TRUE)
cur.spike <- isSpike(sce)
points(var.out$mean[cur.spike], var.out$total[cur.spike], col="red", pch=16)
```

We check the distribution of expression values for the genes with the largest biological components.
This ensures that the variance estimate is not driven by one or two outlier cells (Figure \@ref(fig:hvgvioplot416b)).

```{r hvgvioplot416b, fig.cap="Violin plots of normalized log-expression values for the top 10 genes with the largest biological components in the 416B dataset. Each point represents the log-expression value in a single cell."}
chosen.genes <- order(var.out$bio, decreasing=TRUE)[1:10]
plotExpression(sce, features=rownames(var.out)[chosen.genes]) 
```


# Removing the batch effect

As previously mentioned, the data were collected on two plates.
Small uncontrollable differences in processing between plates can result in a batch effect, i.e., systematic differences in expression between cells on different plates.
Such differences are not interesting and can be removed by applying the `removeBatchEffect()` function from the `r Biocpkg("limma")` package [@ritchie2015limma].
This removes the effect of the plate of origin while accounting for the (interesting) effect of oncogene induction.

`removeBatchEffect()` performs a linear regression and sets the coefficients corresponding to the blocking factors to zero.
This is effective provided that the population composition within each batch is known (and supplied as `design=`) or identical across batches.
Such an assumption is reasonable for this dataset, involving a homogeneous cell line population on both plates.
However, in most scRNA-seq applications, the factors of variation are not identical across batches and not known in advance.


```{r}
library(limma)
assay(sce, "corrected") <- removeBatchEffect(logcounts(sce), 
    design=model.matrix(~sce$Oncogene), batch=sce$Plate)
assayNames(sce)
```

Manual batch correction is necessary for downstream procedures that are not model-based, e.g., clustering and most forms of dimensionality reduction.
However, if an analysis method can accept a design matrix, blocking on nuisance factors in the design matrix is preferable to using `removeBatchEffect()`.
This is because the latter does not account for the loss of residual degrees of freedom, nor the uncertainty of estimation of the blocking factor terms.

# Save object for later use

Once the preprocessing is completed, it is often useful to save the `SingleCellExperiment` object to file with the `saveRDS` function.
The object can then be easily restored into new R sessions using the `readRDS` function.
This allows further work to be conducted without having to repeat all of the processing steps described above.

```{r}
saveRDS(file="416B_postQC.rds", sce)
```

***
# Exercises

For these exercises, use the pre-QC `SingleCellExperiment` object that we saved above.

```{r}
sce <- readRDS("416B_preQC.rds")
```

## How many cells?

How many cells were on each plate? How many cells were induced? Create a 2x2 table that displays the number of cells in each plate and phenotype combination.

```{r}
# your code here

```

## What proportion of measurements are zero?

```{r}
# your code here

```

## How many mitochondrial genes?

How many mitochondrial genes are present?

```{r}
# your code here

```

## How many cells were removed?

How many outlier cells were removed in total?

```{r}
# your code here

```

## Average expressed count versus probability of expression

Examine the relationship between the average read count and the number of expressed cells, but taking the average count only among *expressed* cells (with count > 0). Hint: manually calculate the average count among nonzero cells, and plot against `num.cells` as before. 

```{r}
# your code here

```

## Compare size factors

For this exercise, it is helpful to use the post-QC `SingleCellExperiment` object.

```{r}
sce <- readRDS("416B_postQC.rds")
```

Compute DESeq2 size factors using `estimateSizeFactorsForMatrix`.

```{r}
# your code here

```

How do they compare with the pooled size factor estimates? 
```{r}
# your code here

```

How many genes are used in the calcualtion of DESeq2 size factors? 

```{r}
# your code here

```

## Count-depth relationship

Pick a gene with high average expression. Plot the relationship between sequencing depth (size factors) and count of that gene.

```{r}
# your code here

```

## Highly variable genes without transformation

What would be the result of selecting the most highly variable genes without log transformation? Plot the distribution of expression values for the genes with the largest (untransformed) variance.

```{r}
# your code here

```

***

# Session Info

```{r}
sessionInfo()
```

# References
